### PR TITLE
Fix http request timeout usage in HTML5 build

### LINKED ISSUE #8235: http.request ignore options in html5 platform
**Describe the bug (REQUIRED)**
In http.request you can pass options table with timeout and other options.
In html5 build you ignore all options except timeout.

1)Add comment in docs for options that ignored in html5.
https://defold.com/ref/http/#http.request:url-method-callback-[headers]-[post_data]-[options]

2)You get timeout option from table but not use it.Looks like bug
https://github.com/defold/defold/blob/750368cbae6e010f68f8f4c71c244c59f1e0bda2/engine/script/src/script_http_js.cpp#L160

**Expected behavior (REQUIRED)**
http.request use options

**Defold version (REQUIRED):**
  - Version 1.6.1

**Platforms (REQUIRED):**
 - Platforms: HTML5

### PR DESCRIPTION
**Pull request by ekharkunov**
* Fix timeout option usage in http module for HTML5 implementation. 
* Update `http` module documentation.
* Wrap `HttpRequestAsync` arguments with struct.
* Fix possible memory leak.

Fixes #8235.
## PR checklist

* [ ] Code
	* [ ] Add engine and/or editor unit tests.
	* [ ] New and changed code follows the overall code style of existing code
	* [ ] Add comments where needed
* [ ] Documentation
	* [ ] Make sure that API documentation is updated in code comments
	* [ ] Make sure that manuals are updated (in github.com/defold/doc)
* [ ] Prepare pull request and affected issue for automatic release notes generator
	* [ ] Pull request - Write a message that explains what this pull request does. What was the problem? How was it solved? What are the changes to APIs or the new APIs introduced? This message will be used in the generated release notes. Make sure it is well written and understandable for a user of Defold.
	* [ ] Pull request - Write a pull request title that in a sentence summarises what the pull request does. Do not include "Issue-1234 ..." in the title. This text will be used in the generated release notes.
	* [ ] Pull request - Link the pull request to the issue(s) it is closing. Use on of the [approved closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).
	* [ ] Affected issue - Assign the issue to a project. Do not assign the pull request to a project if there is an issue which the pull request closes.
	* [ ] Affected issue - Assign the "breaking change" label to the issue if introducing a breaking change.
	* [ ] Affected issue - Assign the "skip release notes" is the issue should not be included in the generated release notes.

----------

Example of a well written PR description:

1. Start with the user facing changes. This will end up in the release notes.
1. Add one of the GitHub approved closing keywords
1. Optionally also add the technical changes made. This is information that might help the reviewer. It will not show up in the release notes. Technical changes are identified by a line starting with one of these:
   1. `### Technical changes` 
   1. `Technical changes:`
   2. `Technical notes:`

```
There was a anomaly in the carbon chroniton propeller, introduced in version 8.10.2. This fix will make sure to reset the phaser collector on application startup.

Fixes #1234

### Technical changes
* Pay special attention to line 23 of phaser_collector.clj as it contains some interesting optimizations
* The propeller code was not taking into account a negative phase.
```

### CODE REVIEW

File: engine/script/src/script_http_js.cpp [hunk 1]
```diff
@@ -53,12 +53,24 @@ namespace dmScript
...
     extern "C" void dmScriptHttpRequestAsync(const char* method, const char* url, const char* headers, void* arg, OnLoad onload, OnError onerror, const void* send_data, int send_data_length, int timeout);
 
-    void HttpRequestAsync(const char* method, const char* url, const char* headers, void *arg, OnLoad ol, OnError oe, const void* send_data, int send_data_length)
+    void HttpRequestAsync(const RequestParams& request_params, OnLoad ol, OnError oe)
     {
-        int timeout  = (int) g_Timeout;
-        dmScriptHttpRequestAsync(method, url, headers, arg, ol, oe, send_data, send_data_length, timeout);
+        dmScriptHttpRequestAsync(request_params.m_Method, request_params.m_Url, request_params.m_Headers,
+            request_params.m_Args, ol, oe, request_params.m_SendData, request_params.m_SendDataLength, request_params.m_RequestTimeout);
+        free((void*) request_params.m_SendData);
```
Comment (ekharkunov) [inline engine/script/src/script_http_js.cpp:73]: Free memory that was allocated here https://github.com/defold/defold/pull/8371/files#diff-8df27bb2adf2e376b26a55b64dd6d84fbec204dc69af00239bb14b3c6d3182d2R167

Comment (JCash) [reply engine/script/src/script_http_js.cpp:73]: If not a problem, I think we should release the memory on the caller site, after the HttpRequestAsync call.
That way the ownership/lifetime is clearer for the allocated memory.

**Review by JCash — CHANGES_REQUESTED** [review]

File: engine/script/src/script_http_js.cpp (hunk 1 above)
Comment (ekharkunov) [reply engine/script/src/script_http_js.cpp:73]: We don't store `request_params` after making `dmScriptHttpRequestAsync` call.
So do you want to store at least `request_params.m_SendData` in `RequestContext` struct and free it inside `SendResponse`?

Comment (ekharkunov) [reply engine/script/src/script_http_js.cpp:73]: @JCash Please, confirm am I understand purpose correctly.

Comment (JCash) [reply engine/script/src/script_http_js.cpp:73]: What I meant was that, if it's ok to free the memory here, then the dmScriptHttpRequestAsync() has already copied the data safely?
And then it makes less sense to free the memory here, as opposed to outside of this function, at the caller site.

Comment (ekharkunov) [reply engine/script/src/script_http_js.cpp:73]: Ahhh... Sorry.
When I tested changes I also thought about it. At js side it has direct access to the memory via `HEAPU8.subarray(send_data, send_data + send_data_length)`. But it passed to `XMLHttpRequest.send`. I'm npt sure that XMLHttpRequest take the ownership of the data. I'll look for some documentation regarding that.
At least during my test there were no issues with memory access. I'll recheck.

File: engine/script/src/script_http.cpp [hunk 2]
```diff
@@ -84,9 +84,9 @@ namespace dmScript
      * @param [options] [type:table] optional table with request parameters. Supported entries:
      *
      * - [type:number] `timeout`: timeout in seconds
-     * - [type:string] `path`: path on disc where to download the file. Only overwrites the path if status is 200
-     * - [type:boolean] `ignore_cache`: don't return cached data if we get a 304
-     * - [type:boolean] `chunked_transfer`: use chunked transfer encoding for https requests larger than 16kb. Defaults to true.
+     * - [type:string] `path`: path on disc where to download the file. Only overwrites the path if status is 200. [icon:attention] Not avaialable in HTML5 build
+     * - [type:boolean] `ignore_cache`: don't return cached data if we get a 304. [icon:attention] Not avaialable in HTML5 build
+     * - [type:boolean] `chunked_transfer`: use chunked transfer encoding for https requests larger than 16kb. Defaults to true. [icon:attention] Not avaialable in HTML5 build
```
[the remark this answers is not in the document]
Comment (ekharkunov) [reply engine/script/src/script_http.cpp]: Fixed!

File: engine/script/src/script_http_js.cpp (hunk 1 above)
Comment (ekharkunov) [reply engine/script/src/script_http_js.cpp]: I cannot find any information confirms that javascript part take ownership of memory block. So I've added void* to `RequestContext` structure and free allocated for `m_SendData` memory in request callback.

**Comment by ekharkunov** [comment]
Also branch was rebased to `dev` HEAD.

**Review by JCash — APPROVED** [review]

**Pull request merged**